lib/sqlalchemy/sql/elements.py

```diff
@@ -4472,15 +4472,21 @@ def over(
             rows=rows,
         )
 
-    def self_group(self, against=None):
+    def self_group(
+        self, against: Optional[OperatorType] = None
+    ) -> Union[Self, Grouping[_T]]:
         if operators.is_precedent(operators.filter_op, against):
             return Grouping(self)
         else:
             return self
 
-    @util.memoized_property
-    def type(self):
-        return self.func.type
+    type: TypeEngine[_T]
+
+    if not TYPE_CHECKING:
+        # A001 is a false-positive here, type is a class member
+        @util.memoized_property
+        def type(self) -> TypeEngine[_T]:  # noqa: A001
```

Comment:
**Federico Caselli** (CaselIT) wrote:

strange that this is needed, what issue does the type checker has?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

Reply:
**Federico Caselli** (CaselIT) wrote:

by "this" I mean the `if not Type_ckeching` block

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

Reply:
**Michael Bayer** (zzzeek) wrote:

probably needs a different memoized decorator, ro

edit: this is also on the superclass.  i will try to fix it as this is a much newer mypy

edit: no we definitely need this, many many subclasses have a simple "type: TypeEngine" variable and simple assignment is used.   mypy does not understand the `@memoized_property` thing so we have to conceal it here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

Reply:
**Michael Bayer** (zzzeek) wrote:

we don't actually need the type-checked annotation here though, as the base class has this already

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972